### ISSUE TITLE
Access to ConnectionPool.count should be thread safe

### ISSUE THREAD

**Issue by ya-pulser** [swift-server/async-http-client#318]
ConnectionPool has a collection of `providers` registered with the pool and  access to the collection is guarded by ConnectionPool.lock.

Size of the collection is exposed through `count` accessor without guarding by the lock.
https://github.com/swift-server/async-http-client/blob/1.2.2/Sources/AsyncHTTPClient/ConnectionPool.swift#L112

That leads to warnings in thread safety analyser.

To reproduce run with thread safety analyzer:

```
docker run --rm -v $(pwd):/src -w /src swift:5.3.1 swift test -c release --sanitize=thread --enable-test-discovery
```

Analysis of output shows that this `count` accessor is main source of thread safety warnings.

**Comment by ya-pulser** [comment]
PR #319 as a proposed solution

**Comment by Lukasa** [comment]
I believe we’re done here: see https://github.com/swift-server/async-http-client/pull/319#discussion_r530515030 for more reasoning.

**Comment by Lukasa** [comment]
@weissi has a good reference for a TSAN issue in 5.3 that manifests this way.

**Comment by weissi** [comment]
@fabianfett I think we can close this?

**Comment by fabianfett** [comment]
I'll add it to the http/2 milestone and close it as soon as the connection pool has landed.

**Comment by fabianfett** [comment]
Fixed in #319 and removed from an active code path in #427.
